withData ignores assertSoftly configuration

I'm using the latest version of Kotest. It doesn't matter where I configure that I want to use soft assertions by default. I'm doing it in the test case in the minimal reproducer, but the same effect applies if it's configured globally via a config class or the properties.

```kotlin
package com.fleshgrinder.kotest

import io.kotest.core.spec.style.StringSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe

class WithDataSoftly : StringSpec({
    assertSoftly = true

    withData(Pair("a", "b"), Pair("c", "d")) {
        it.first shouldBe "x"
        it.second shouldBe "y"
    }
})
```

```
Expected :"x"
Actual   :"a"
<Click to see difference>

[…]

Expected :"x"
Actual   :"c"
<Click to see difference>
```

Wrapping the assertions myself in `assertSoftly` fixes the issue.

```kotlin
package com.fleshgrinder.kotest

import io.kotest.assertions.assertSoftly
import io.kotest.core.spec.style.StringSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe

class WithDataSoftly : StringSpec({
    withData(Pair("a", "b"), Pair("c", "d")) {
        assertSoftly {
            it.first shouldBe "x"
            it.second shouldBe "y"
        }
    }
})
```

```
io.kotest.assertions.MultiAssertionError: The following 2 assertions failed:
1) expected:<"x"> but was:<"a">
   at com.fleshgrinder.kotest.WithDataSoftly$1$1.invokeSuspend(WithDataSoftly.kt:11)
2) expected:<"y"> but was:<"b">
   at com.fleshgrinder.kotest.WithDataSoftly$1$1.invokeSuspend(WithDataSoftly.kt:12)

[…]

io.kotest.assertions.MultiAssertionError: The following 2 assertions failed:
1) expected:<"x"> but was:<"c">
   at com.fleshgrinder.kotest.WithDataSoftly$1$1.invokeSuspend(WithDataSoftly.kt:11)
2) expected:<"y"> but was:<"d">
   at com.fleshgrinder.kotest.WithDataSoftly$1$1.invokeSuspend(WithDataSoftly.kt:12)
```

It does however break the `<Click to see difference>` but that's a different story. 😉

--- a/kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/AssertSoftlyTestLevelTest.kt
+++ b/kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/matchers/AssertSoftlyTestLevelTest.kt
@@ -1,0 +1,42 @@
+package com.sksamuel.kotest.matchers
+
+import io.kotest.core.config.ProjectConfiguration
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.engine.TestEngineLauncher
+import io.kotest.engine.listener.CollectingTestEngineListener
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldContain
+
+class AssertSoftlyTestLevelTest : FunSpec() {
+   init {
+      assertSoftly = true
+
+      context("test level assertSoftly should work at the container level") {
+         val listener = CollectingTestEngineListener()
+         TestEngineLauncher(listener)
+            .withClasses(AssertSoftlyAtTestLevel::class)
+            .withConfiguration(ProjectConfiguration())
+            .launch()
+         listener.tests.size shouldBe 2
+         listener.errors shouldBe true
+         listener.result("test level assertSoftly should work at the container level")!!.errorOrNull!!.message shouldContain "The following 2 assertions failed"
+         listener.result("test level assertSoftly should work at the test level")!!.errorOrNull!!.message shouldContain "The following 2 assertions failed"
+      }
+   }
+}
+
+private class AssertSoftlyAtTestLevel : FunSpec() {
+   init {
+      assertSoftly = true
+
+      context("test level assertSoftly should work at the container level") {
+         1 shouldBe 2
+         1 shouldBe 3
+      }
+
+      test("test level assertSoftly should work at the test level") {
+         1 shouldBe 2
+         1 shouldBe 3
+      }
+   }
+}

--- a/kotest-assertions/kotest-assertions-shared/src/jvmTest/kotlin/io/kotest/assertions/AssertSoftlyTests.kt
+++ b/kotest-assertions/kotest-assertions-shared/src/jvmTest/kotlin/io/kotest/assertions/AssertSoftlyTests.kt
@@ -5,7 +5,6 @@ import io.kotest.core.spec.style.FunSpec
 import io.kotest.matchers.collections.shouldHaveSize
 import io.kotest.matchers.shouldBe
 import io.kotest.matchers.string.shouldContainInOrder
-import io.kotest.matchers.string.shouldNotContain
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.delay
 import kotlinx.coroutines.withContext

--- a/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/core/TestConfiguration.kt
+++ b/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/core/TestConfiguration.kt
@@ -67,6 +67,9 @@ abstract class TestConfiguration {
     */
    var assertions: AssertionMode? = null
 
+   /**
+    * Whether soft assertion mode should be applied for all tests in the spec.
+    */
    var assertSoftly: Boolean? = null
 
    /**

--- a/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/SoftAssertInterceptor.kt
+++ b/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/SoftAssertInterceptor.kt
@@ -3,17 +3,15 @@ package io.kotest.engine.test.interceptors
 import io.kotest.assertions.assertSoftly
 import io.kotest.assertions.collectiveError
 import io.kotest.assertions.errorCollector
+import io.kotest.core.Logger
 import io.kotest.core.test.TestCase
 import io.kotest.core.test.TestResult
 import io.kotest.core.test.TestScope
-import io.kotest.core.test.TestType
-import io.kotest.core.Logger
 
 /**
- * Executes the test with assertSoftly if [assertSoftly] is enabled for this test
- * and if the [TestCase] is a [TestType.Test].
+ * Executes the test with assertSoftly if [assertSoftly] is enabled for this test or container.
  */
-internal class SoftAssertInterceptor() : TestExecutionInterceptor {
+internal class SoftAssertInterceptor : TestExecutionInterceptor {
 
    private val logger = Logger(SoftAssertInterceptor::class)
 
@@ -23,7 +21,6 @@ internal class SoftAssertInterceptor() : TestExecutionInterceptor {
       test: NextTestExecutionInterceptor
    ): TestResult {
 
-      if (testCase.type != TestType.Test) return test(testCase, scope)
       if (!testCase.config.assertSoftly) return test(testCase, scope)
 
       logger.log { Pair(testCase.name.testName, "Invoking test with soft assert") }

